### PR TITLE
Engine observer clear search terms

### PR DESCRIPTION
**Pull request by codrut-topliceanu**
For [#7632](https://github.com/mozilla-mobile/fenix/issues/7632)

---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks.
- [x] **Tests**: This PR includes thorough tests.
- [x] **Changelog**: This PR does not need one.
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md).

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -133,6 +133,7 @@ class GeckoEngineSession(
      */
     override fun goBack() {
         geckoSession.goBack()
+        notifyObservers { onNavigateBack() }
```
Comment (pocmo) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Can you add a small test for the `engine-gecko` components that verifies that calling `goBack()` will invoke the observer?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 2]
```diff
@@ -42,6 +42,7 @@ abstract class EngineSession(
         fun onTrackerBlockingEnabledChange(enabled: Boolean) = Unit
         fun onTrackerBlocked(tracker: Tracker) = Unit
         fun onTrackerLoaded(tracker: Tracker) = Unit
+        fun onNavigateBack() = Unit
```
Comment (pocmo) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: Can we implement this for `browser-engine-system` too? :)

File: components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt [hunk 3]
```diff
@@ -39,6 +39,10 @@ internal class EngineObserver(
     private val store: BrowserStore? = null
 ) : EngineSession.Observer {
 
+    override fun onNavigateBack() {
+        if (session.canGoBack) session.searchTerms = ""
```
Comment (pocmo) [inline components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt]: I think the `canGoBack()` check may need to happen in the engine implementation. Now we are calling `canGoBack` after calling `goBack()` and there's no guarantee whether this navigation has already happened or not..?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 2 above)
Comment (codrut-topliceanu) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: I've implemented this in SystemEngineSession as well.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (codrut-topliceanu) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Done :D

File: components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt (hunk 3 above)
Comment (codrut-topliceanu) [reply components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt]: You are right, thank you. I've moved the check in the engine implementations.

**Comment by csadilek** [comment]
Sorry for the delay @codrut-topliceanu. Would you mind rebasing / addressing the merge conflict that happened in the meantime? Then we can land this.

**Comment by codrut-topliceanu** [comment]
All done :D Thank you!

**Pull request merged**